lib/Sema/CSSimplify.cpp

```diff
@@ -10185,6 +10242,42 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyFixConstraint(
     return recordFix(fix, impact) ? SolutionKind::Error : SolutionKind::Solved;
   }
 
+  case FixKind::TreatArrayLiteralAsDictionary: {
+    ArrayExpr *AE = getAsExpr<ArrayExpr>(fix->getAnchor());
+    assert(AE);
+
+    // If the array was empty, there's nothing to do.
+    if (AE->getNumElements() == 0)
+      return recordFix(fix) ? SolutionKind::Error : SolutionKind::Solved;
+
+    // For arrays with a single element, match it to the dictionary's key type.
+    auto elemTy = isArrayType(type1);
+    SmallVector<Type, 2> optionals;
+    auto dictTy = type2->lookThroughAllOptionalTypes(optionals);
+    auto valueTy = createTypeVariable(getConstraintLocator(locator),
+                                      TVO_CanBindToLValue |
+                                        TVO_CanBindToNoEscape |
+                                        TVO_CanBindToHole);
+    assignFixedType(valueTy, dictTy);
+
+    ProtocolDecl *dictionaryProto =
+        Context.getProtocol(KnownProtocolKind::ExpressibleByDictionaryLiteral);
+    auto keyAssocTy = dictionaryProto->getAssociatedType(Context.Id_Key);
+    auto dictionaryKeyTy = DependentMemberType::get(valueTy, keyAssocTy);
+
+    ConstraintKind kind = isDictionaryType(dictTy)
+        ? ConstraintKind::Conversion
+        : ConstraintKind::Equal;
+
+    ConstraintLocator *elemLoc = getConstraintLocator(AE->getElement(0));
+    auto result = matchTypes(*elemTy, dictionaryKeyTy, kind, subflags, elemLoc);
+    if (result.isSuccess()) {
+        if (recordFix(fix, optionals.size() + 1))
```

Comment:
I'd suggest you first try to record and if that succeeds then do matching and check the result since there is no point of going through the motions here is the solution is already known to be worse than the best one.